File: packages/core/src/schemas/metadata/MetadataApplication.ts
import { ClassProperties, IMetadataSchemaCollection } from "@typia/interface";

import { MetadataComponents } from "./MetadataComponents";
import { MetadataSchema } from "./MetadataSchema";

export class MetadataApplication {
  public readonly schemas: MetadataSchema[];
  public readonly components: MetadataComponents;

  private constructor(props: ClassProperties<MetadataApplication>) {
    this.schemas = props.schemas;
    this.components = props.components;
  }

  /** @internal */
  public static create(
    props: ClassProperties<MetadataApplication>,
  ): MetadataApplication {
    return new MetadataApplication(props);
  }

  public static from(app: IMetadataSchemaCollection): MetadataApplication {
    const components: MetadataComponents = MetadataComponents.from(
      app.components,
    );
    const metadatas: MetadataSchema[] = app.schemas.map((metadata) =>
      MetadataSchema.from(metadata, components.dictionary),
    );
    return MetadataApplication.create({ schemas: metadatas, components });
  }

  public toJSON(): IMetadataSchemaCollection {
    return {
      schemas: this.schemas.map((s) => s.toJSON()),
      components: this.components.toJSON(),
    };
  }
}

File: packages/core/src/schemas/metadata/MetadataApplication.go
package metadata

type MetadataApplication struct {
	Schemas    []*MetadataSchema
	Components *MetadataComponents
}

type MetadataApplicationProps struct {
	Schemas    []*MetadataSchema
	Components *MetadataComponents
}

func CreateMetadataApplication(props MetadataApplicationProps) *MetadataApplication {
	return &MetadataApplication{
		Schemas:    append([]*MetadataSchema(nil), props.Schemas...),
		Components: props.Components,
	}
}

func MetadataApplicationFrom(app MetadataSchemaCollectionJSON) (*MetadataApplication, error) {
	components, err := MetadataComponentsFrom(app.Components)
	if err != nil {
		return nil, err
	}
	schemas := make([]*MetadataSchema, 0, len(app.Schemas))
	for _, schema := range app.Schemas {
		next, err := MetadataSchemaFrom(schema, components.Dictionary)
		if err != nil {
			return nil, err
		}
		schemas = append(schemas, next)
	}
	return CreateMetadataApplication(MetadataApplicationProps{
		Schemas:    schemas,
		Components: components,
	}), nil
}

func (m *MetadataApplication) ToJSON() MetadataSchemaCollectionJSON {
	schemas := make([]MetadataSchemaJSON, 0, len(m.Schemas))
	for _, schema := range m.Schemas {
		schemas = append(schemas, schema.ToJSON())
	}
	return MetadataSchemaCollectionJSON{
		Schemas:    schemas,
		Components: m.Components.ToJSON(),
	}
}
